File: DP/elevator-rides.py
n, x = map(int, input().split())
arr = [int(y) for y in input().split()]

INF = 10 ** 9

dp = [[INF, INF] for _ in range(1 << n)]
dp[0][0] = 1
dp[0][1] = 0

for mask in range(1, 1<<n):
    for i in range(n):
        if (mask & 1<<i) > 0:
            prev = mask ^ (1<<i)
            numberofElevators, totalCount = dp[prev]
            # if totalCount + arr[i] <= x:
            #     dp[mask][0] = min(numberofElevators, dp[mask][0])
            #     dp[mask][1] = totalCount + arr[i]
            # else:
            #     dp[mask][0] = numberofElevators + 1
            #     dp[mask][1] = min(dp[mask][1], arr[i])

            if totalCount + arr[i] <= x:
                dp[mask] = min(dp[mask], [numberofElevators, totalCount+arr[i]])
            else:
                dp[mask] = min(dp[mask], [numberofElevators+1, arr[i]])

print(dp[(1<<n) - 1][0])

File: DP/elevator-rides.cpp
#include <iostream>
#include <vector>
#include <algorithm>

const int INF = 1e9;

int main() {
    int n, x;
    std::cin >> n >> x;

    std::vector<int> arr(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> arr[i];
    }

    std::vector<std::vector<int>> dp(1 << n, std::vector<int>(2, INF));
    dp[0][0] = 1;
    dp[0][1] = 0;

    for (int mask = 1; mask < (1 << n); ++mask) {
        for (int i = 0; i < n; ++i) {
            if ((mask & (1 << i)) > 0) {
                int prev = mask ^ (1 << i);
                int numberofElevators = dp[prev][0];
                int totalCount = dp[prev][1];

                if (totalCount + arr[i] <= x) {
                    dp[mask] = std::min(dp[mask], std::vector<int>{numberofElevators, totalCount + arr[i]});
                } else {
                    dp[mask] = std::min(dp[mask], std::vector<int>{numberofElevators + 1, arr[i]});
                }
            }
        }
    }

    std::cout << dp[(1 << n) - 1][0] << std::endl;

    return 0;
}
